### PR TITLE
RPC servers use attestation pool

### PR DESCRIPTION
**Pull request by terencechain**
This hooks RPC servers with the new attestation pool functionalities implemented for v0.9.2

Change list:
- Modified attester server to call `SaveUnaggregatedAttestation` to save latest submitted attestation from the validator to the attestation pool
- Modified Proposer server to call `AggregatedAttestation` to retrieve a list of aggregated attestations from pool then filter them based on validability and profitability
- Add Aggregater server to call `UnaggregatedAttestation` to retrieve unaggregated attstations filtered based on slot and committee index then to aggregate them and broadcast them

### CODE REVIEW

File: beacon-chain/rpc/aggregator/server.go [hunk 1]
```diff
@@ -71,7 +79,36 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *pb.Aggregati
 		return nil, status.Errorf(codes.InvalidArgument, "Validator is not an aggregator")
 	}
 
-	// TODO(3865): Broadcast aggregated attestation & proof via the aggregation topic
+	// Retrieve the aggregated attestation from pool
```
Comment (rauljordan) [inline beacon-chain/rpc/aggregator/server.go]: retrieve the unaggreated* right?

File: beacon-chain/rpc/aggregator/server.go [hunk 2]
```diff
@@ -71,7 +79,36 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *pb.Aggregati
...
+		validAtts = append(validAtts, att)
+	}
+
+	// Aggregate the attestations and broadcast them
+	aggregatedAtts, err := helpers.AggregateAttestations(validAtts)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not aggregate attestations: %v", err)
+	}
+	for _, att := range aggregatedAtts {
+		if att.AggregationBits.Count() > 1 {
```
Comment (rauljordan) [inline beacon-chain/rpc/aggregator/server.go]: Maybe we can create a public helper func called IsAggregated in some shared package somewhere

File: beacon-chain/rpc/aggregator/server_test.go [hunk 3]
```diff
@@ -90,3 +99,132 @@ func TestSubmitAggregateAndProof_IsAggregator(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	aggregatedAtts := aggregatorServer.AttPool.AggregatedAttestation()
+	if len(aggregatedAtts) != 0 {
+		t.Errorf("Wanted aggregated attestation 0, got %d", len(aggregatedAtts))
+	}
+}
+
+func GenerateAtt(state *pbp2p.BeaconState, index uint64, privKeys []*bls.SecretKey) *ethpb.Attestation {
```
Comment (rauljordan) [inline beacon-chain/rpc/aggregator/server_test.go]: Why exported? If this is used in various places can we put it in the testutil? There are some utils that generate blocks as well and would be nice to have these in one place

File: beacon-chain/rpc/aggregator/server.go [hunk 4]
```diff
@@ -71,7 +79,36 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *pb.Aggregati
 		return nil, status.Errorf(codes.InvalidArgument, "Validator is not an aggregator")
 	}
 
-	// TODO(3865): Broadcast aggregated attestation & proof via the aggregation topic
+	// Retrieve the aggregated attestation from pool
+	atts := as.AttPool.UnaggregatedAttestations(req.Slot, req.CommitteeIndex)
+
+	// Verify attestations are valid before aggregating and broadcasting them out
```
Comment (rauljordan) [inline beacon-chain/rpc/aggregator/server.go]:
```suggestion
	// Verify attestations are valid before aggregating and broadcasting them out.
```

File: beacon-chain/rpc/aggregator/server.go [hunk 5]
```diff
@@ -71,7 +79,36 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *pb.Aggregati
...
+		if err := blocks.VerifyAttestation(ctx, headState, att); err != nil {
+			if err := as.AttPool.DeleteUnaggregatedAttestation(att); err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not delete invalid attestation: %v", err)
+			}
+			continue
+		}
+		validAtts = append(validAtts, att)
+	}
+
+	// Aggregate the attestations and broadcast them
```
Comment (rauljordan) [inline beacon-chain/rpc/aggregator/server.go]:
```suggestion
	// Aggregate the attestations and broadcast them.
```

File: beacon-chain/rpc/proposer/server.go [hunk 6]
```diff
@@ -356,6 +357,49 @@ func (ps *Server) defaultEth1DataResponse(ctx context.Context, currentHeight *bi
 	}, nil
 }
 
+// This filters the input attestations to return a list of valid attestations to be packaged inside a beacon block.
+func (ps *Server) filterAtts(ctx context.Context, slot uint64, atts []*ethpb.Attestation) ([]*ethpb.Attestation, error) {
```
Comment (rauljordan) [inline beacon-chain/rpc/proposer/server.go]: Naming could be clearer - something like `filterAttestationsReadyForInclusion`. Would rather be more verbose in naming these RPC helper funcs for future reference.

File: beacon-chain/rpc/aggregator/server.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/rpc/aggregator/server.go]: Nice catch

File: beacon-chain/rpc/aggregator/server_test.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/rpc/aggregator/server_test.go]: You are right. It doesn't have to be exported

File: beacon-chain/core/helpers/attestation.go [hunk 7]
```diff
@@ -165,3 +165,12 @@ func AggregateSignature(attestations []*ethpb.Attestation) (*bls.Signature, erro
 	}
 	return aggregateSignatures(sigs), nil
 }
+
+// IsAggregated returns true if the attestation is an aggregated attestation,
+// false otherwise.
+func IsAggregated(attestation *ethpb.Attestation) bool {
```
Comment (rauljordan) [inline beacon-chain/core/helpers/attestation.go]: return attestation.AggregationBits.Count() > 1

**Comment by prestonvanloon** [comment]
This PR failed E2E tests. Please fix as soon as possible.

**Pull request merged**